File: src/Starter/Views/SearchResult.axaml.fs
namespace Starter.Controls

open Starter.ViewModels
open Starter.SearchEngine

open Avalonia.Controls
open Avalonia.Markup.Xaml

type SearchResultControl() as this =
    inherit UserControl()

    do this.InitializeComponent()

    member this.InitializeComponent() =
        AvaloniaXamlLoader.Load this

File: src/Starter/Controls/SearchResult.axaml.cs
using Avalonia.Controls;
using Avalonia.Controls.Templates;
using Starter.Features;
using Starter.SearchEngine;

namespace Starter.Controls;

public partial class SearchResult : UserControl
{
    public SearchResult() => InitializeComponent();
}

public class SearchResultDataTemplate : IRecyclingDataTemplate
{
    public bool Match(object? data) => data is null || data is SearchResultData;

    public Control? Build(object? param, Control? existing)
    {
        if (param is not SearchResultData data) return null;

        return data.SearchResult is IControlSearchResult
            ? existing as ControlSearchResult ?? new ControlSearchResult()
            : existing ?? new SearchResult();
    }

    public Control? Build(object? param) => Build(param, null);
}
